File: sdk/ts/src/provenance.ts
/**
 * @module provenance
 * @package @hop-top/kit
 *
 * Factor 11 — Provenance.
 *
 * Attaches origin metadata to any data payload for JSON output,
 * enabling consumers to verify source, timing, and method.
 */

export interface Provenance {
  source: string;
  /** ISO 8601 timestamp */
  timestamp: string;
  method: string;
}

/**
 * Wraps `data` with provenance metadata under `_meta`.
 */
export function withProvenance<T>(
  data: T,
  p: Provenance,
): { data: T; _meta: Provenance } {
  return { data, _meta: p };
}

File: sdk/py/hop_top_kit/provenance.py
"""provenance — data provenance metadata (12-factor AI CLI, Factor 11).

Wraps output data with _meta containing source, timestamp, and method.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any


@dataclass
class Provenance:
    source: str
    timestamp: str  # ISO 8601
    method: str


def with_provenance(data: Any, p: Provenance) -> dict:
    """Wrap data with provenance metadata."""
    return {"data": data, "_meta": asdict(p)}
